### ISSUE TITLE
HTML assets not loading due to CORS configuration and Content Policy issues

### ISSUE THREAD

**Issue by codewithfeeling** [hicetnunc2000/hicetnunc#250]
I minted an NFT which includes webfonts. The files exist on Cloudflare and the HTML directory is completely flat in the correct structure, same as the template, but the fonts are not loading in hicetnunc.xyz due to CORS issues:

````
cloudflare-ipfs.com/:1 Access to font at 'https://cloudflare-ipfs.com/ipfs/QmZbTQFJ1TWMLWfrKGe2kQvo71q12VMKCcd4Tvb4AcyVpV/SairaCondensed-Regular.ttf' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
SairaCondensed-Regular.ttf:1 Failed to load resource: net::ERR_FAILED

cloudflare-ipfs.com/:1 Access to font at 'https://cloudflare-ipfs.com/ipfs/QmZbTQFJ1TWMLWfrKGe2kQvo71q12VMKCcd4Tvb4AcyVpV/FredokaOne-Regular.ttf' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
FredokaOne-Regular.ttf:1 Failed to load resource: net::ERR_FAILED
````

The mobile version is much worse - the piece will not render at all, failing with `Content-Policy` errors on the scripts and styles:

![image](https://user-images.githubusercontent.com/262516/112437981-79273200-8d3f-11eb-8824-d9fce3c92b30.png)

````
[Error] Refused to load https://cloudflare-ipfs.com/ipfs/QmZbTQFJ1TWMLWfrKGe2kQvo71q12VMKCcd4Tvb4AcyVpV/style.css because it appears in neither the style-src directive nor the default-src directive of the Content Security Policy.
[Error] Refused to load https://cloudflare-ipfs.com/ipfs/QmZbTQFJ1TWMLWfrKGe2kQvo71q12VMKCcd4Tvb4AcyVpV/app.js because it appears in neither the script-src directive nor the default-src directive of the Content Security Policy.
````

The piece works perfectly if loaded locally or [on my own server](https://theveltrons.com/swoosh2/).

**Comment by loackme** [comment]
I got the same problem with my first html piece.
https://www.hicetnunc.xyz/objkt/12655
It works fine on Chrome but I get similar error messages in Safari.

`[Error] Refused to load https://cloudflare-ipfs.com/ipfs/QmUXiq51iFh9G8eAy2aM3e2go4nyfQ46dDWm4XgrTCsGVn/style.css because it appears in neither the style-src directive nor the default-src directive of the Content Security Policy.
[Error] Refused to load https://cloudflare-ipfs.com/ipfs/QmUXiq51iFh9G8eAy2aM3e2go4nyfQ46dDWm4XgrTCsGVn/p5.min.js because it appears in neither the script-src directive nor the default-src directive of the Content Security Policy.
[Error] Refused to load https://cloudflare-ipfs.com/ipfs/QmUXiq51iFh9G8eAy2aM3e2go4nyfQ46dDWm4XgrTCsGVn/script.js because it appears in neither the script-src directive nor the default-src directive of the Content Security Policy.`

**Comment by klevron** [comment]
Hello, I have same problem here for js assets : https://www.hicetnunc.xyz/objkt/13200

![image](https://user-images.githubusercontent.com/4560725/112550619-f1cdd300-8dbf-11eb-8e0d-67a146f228d3.png)

**Comment by piducancore** [comment]
same issue with https://www.hicetnunc.xyz/objkt/13860

**Comment by pichiste** [comment]
Hi all, there seem to be two different things going on here:

The Content Security Policy problem is a known issue specific to Safari and is currently being worked on.

@codewithfeeling I did a simple test with loading a font file and it works:
https://ipfs.io/ipfs/QmecpPXMDLsX37119uQr5XPcgRmbhpfsnVjXHX46PpxMbe/
It looks like you're using React, but can't debug from the bundle. Can send me a stripped down version of your source that just displays some text?

A proper preview for HTML is in the works to catch these sorts of issues before minting.

**Comment by codewithfeeling** [comment]
@pichiste Thanks for looking at this. I'm not using React - it's just vanilla HTML and JavaScript. You are right that loading the font file directly works, but this is not the problem of course. The issue is that when trying to load from hicetnunc.xyz, the font does not load due to CORS issues with CloudFlare. I think you will either need to update the policy on CloudFlare (if that is possible - I had to do this with Amazon CloudFront a while ago) or send some additional headers when making requests.

Here is a ZIP of source code for a minimal reproducible test case:
[Source](https://github.com/hicetnunc2000/hicetnunc/files/6215417/dist.zip)

You should see this:

![image](https://user-images.githubusercontent.com/262516/112716218-f340ed80-8edc-11eb-803d-3618a3d15303.png)

Demo shown working on a server with correct fonts [here](https://mixless.com/hicetnunc/font-cors-issue/)

The Content Security Policy problem is not just Safari. I don't use Safari. I have been unable to get any HTML to work on my phone at all on iOS Brave Mobile Browser or Firefox Mobile browser either. In each case the errors are the same - the Content Policy error. They mostly work fine on Safari desktop, but fail on all mobile browsers on iOS.

**Comment by pichiste** [comment]
@codewithfeeling Thanks for sending over the example and the extra info. I will take a look at it as soon as I get the chance.

Re: Safari, I wasn't referring to CORS issue, but rather the "Refused to load" error that others posted. That's what I meant by two different issues.

**Comment by pichiste** [comment]
I should add, re: Brave, I believe all/most browsers on iOS are using webkit via WKWebView under the hood, so same issue occurs (but don't quote me on that!). I guess I should rephrase, it's more generally a Webkit issue not Safari

**Comment by codewithfeeling** [comment]
Hi @pichiste - that makes sense. I assume it is the same error and something to do with WebKit under the hood. Once the Content Policy is ironed out I'm pretty sure it will start to work... hope you can get it sorted. Thanks for all the great work so far.

**Comment by luispintofraga** [comment]
Hello, I also have problems loading images and sound: https://www.hicetnunc.xyz/objkt/13865

>Access to XMLHttpRequest at 'https://cloudflare-ipfs.com/ipfs/QmbHk6EhKjiDNo5qw8GE6JdjpC1jnhQwJGfVpJm1RsT71a/assets/Synchrotron_-_07092019,_15.46.mp3' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

Chrome/Firefox Mac - visuals ok (meaning separate js files are being properly linked), media fails 
Safari Mac - doesn't even render anything.

I did test this thoroughly in a local webserver to make sure everything was in the zip bundle.

UPDATE: after some discord chats, my issue seems to (also?) be caused by my use of subdirectories, i.e. the "assets" folder. Then again, my piece was pretty much a direct export from cables.gl, and I assume many tools will also export outputs with folders in it. So eventually more creators will also bump into this.

Ideas for "developer experience" mitigations:
- Better documentation (image appears here, should be square, can’t have this, can’t connect there, how to cleanup bad mints)
- A playground / testing sandbox
- An upload zip-verifier (almost like a linter)

**Comment by pichiste** [comment]
@codewithfeeling you were right about the Cloudflare headers, thanks for the suggestion. We can't modify these, but are discussing some solutions (perhaps a proxy). Open to ideas and/or contributions!

Hi @luispintofraga, thanks for reporting. The directory structure shouldn't matter, it seems to just be the same Cloudflare headers mentioned above. The Webkit/Safari issue is known and also being worked on.

Hang in there, this is still an experimental feature that we're working to get stable soon.

**Comment by codewithfeeling** [comment]
Hey @pichiste - thanks for all the hard work. You guys are doing a great job.

**Comment by pichiste** [comment]
Hi all, a fix for the CSP error ("Refused to load...") on webkit browsers has been deployed. 

It's also now possible to preview your HTML NFT before minting, which will let you catch problems like this before minting. 

If any of you could give it a try (no need to mint) and see if the errors still occur in the preview, that would help us track down any remaining issues (i.e. CORS issue). Thanks!

**Comment by codewithfeeling** [comment]
I’m still seeing this for my HTML piece https://www.hicetnunc.xyz/objkt/12381 on all iOS browsers. 

![6490FC1F-D013-429E-91C3-CC3CC9FF0CFA](https://user-images.githubusercontent.com/262516/112882730-3e1d5980-90c5-11eb-8b72-049229beec85.png)

**Comment by pichiste** [comment]
@codewithfeeling The fix will only apply to new pieces unfortunately. Would you mind trying to upload the zip on the mint page to preview (without minting), to see if the errors are gone?

**Comment by pichiste** [comment]
@codewithfeeling The zip you sent seems to work ok:

![Screen Shot 2021-03-29 at 8 41 32 PM](https://user-images.githubusercontent.com/111979/112884034-4c707300-90cf-11eb-9ded-0b0cff92b9d9.png)

**Comment by codewithfeeling** [comment]
Sure - will do that after dinner. Herculean work guys!!!

**Comment by mikedotalmond** [comment]
Not sure if this should be a separate issue, but regarding the Content-Security-Policy tag that gets injected into the index.html -- would it be possible to modify it allow local loading of `data:` and `blob:` types like:

`<meta http-equiv="Content-Security-Policy" content="default-src data: blob: 'self' 'unsafe-inline';">`

I have a project that loads base64 encoded images using `data:` and audio as a binary `blob:` - in my testing I needed to update the CSP meta tag to include those in order to get it to work.

I'm not sure if this would introduce any security issues... but presumably if it's only allowing local sources and the embedded code can't call out anywhere, it should be safe?

**Comment by codewithfeeling** [comment]
My first attempt to embed web fonts was to base64 encode them but that failed. I would LOVE the data: type to be made available - then I could just embed web fonts that way with no additional call to the server.

**Comment by luispintofraga** [comment]
The zip preview is great! Great work.

I confirm I can now see texture being loaded from a /assets folder that was previously failing.
But sound still doesn't play AND the preview doesn't seem to be interactive (the previously minted piece was).
Didn't mint anything tho.

Can share my zip if you need. Or its objtk 13865 if you get the info from there.

**Comment by codewithfeeling** [comment]
@pichiste I'm delighted to say that the ZIP upload direct from my mobile shows the correct preview (on Brave browser), although there is no interactivity when previewing which was a little confusing. I'm just sorting out the new version and then will mint it to check. Looks like fonts will load as well as it all working on mobile... amazing work guys.

**Comment by codewithfeeling** [comment]
I just got this error trying to get a preview on desktop:

````
Uncaught (in promise) DOMException: Failed to register a ServiceWorker for scope ('https://hicetnunc2000.github.io/hicetnunc/gh-pages/html-preview/') with script ('https://hicetnunc2000.github.io/hicetnunc/gh-pages/html-preview/sw.js'): The user denied permission to use Service Worker.
````

**Comment by codewithfeeling** [comment]
Fonts are still not working on the desktop site: https://www.hicetnunc.xyz/objkt/16922 - same errors as before:

````
Access to font at 'https://cloudflare-ipfs.com/ipfs/QmQYcfRJUSWSRdppWaKN58VF1KQMX1xCBb5aJoaNdAiFzx/FredokaOne-Regular.ttf' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
FredokaOne-Regular.ttf:1 Failed to load resource: net::ERR_FAILED
?creator=tz1YJvMiZyXnzvV9pxtAiuCFvaG7XoBZhbUQ&viewer=tz1YJvMiZyXnzvV9pxtAiuCFvaG7XoBZhbUQ:1 Access to font at 'https://cloudflare-ipfs.com/ipfs/QmQYcfRJUSWSRdppWaKN58VF1KQMX1xCBb5aJoaNdAiFzx/SairaCondensed-Regular.ttf' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
SairaCondensed-Regular.ttf:1 Failed to load resource: net::ERR_FAILED
````

but they DO load correctly on mobile!

**Comment by klevron** [comment]
Hello, thanks for the hard work but I have still stange issues.
I tried to preview multiple times the same zip, it works only once, had the following error most of the time : 

![image](https://user-images.githubusercontent.com/4560725/112971982-0e676380-9150-11eb-8a41-1861192a48b7.png)

**Comment by pichiste** [comment]
Hi @klevron what browser are you using? And what do you see in the web console? Thanks

**Comment by pichiste** [comment]
@klevron would you mind opening a new issue about this? Different than the one above

**Comment by klevron** [comment]
The `missing index.html` error is with chrome, just tried with firefox and the preview was ok, but still have the cors problem :
![image](https://user-images.githubusercontent.com/4560725/113016910-38844a00-917f-11eb-8ce4-a20e73f2f8ef.png)

https://www.hicetnunc.xyz/objkt/17506

**Comment by pichiste** [comment]
Hi all, fix is in! Can you check your pieces and let me know if you still have issues? Thanks for your input

**Comment by pichiste** [comment]
> @pichiste I'm delighted to say that the ZIP upload direct from my mobile shows the correct preview (on Brave browser), although there is no interactivity when previewing which was a little confusing. I'm just sorting out the new version and then will mint it to check. Looks like fonts will load as well as it all working on mobile... amazing work guys.

Ah I think the interactivity thing still needs to be fixed, thanks for catching that. I'll open a new ticket

**Comment by luispintofraga** [comment]
Thanks for the quick turnaround!

**Comment by codewithfeeling** [comment]
The preview is still not working for me at all. Blank screen with this error:

````
Uncaught (in promise) DOMException: Failed to register a ServiceWorker for scope ('https://hicetnunc2000.github.io/hicetnunc/gh-pages/html-preview/') with script ('https://hicetnunc2000.github.io/hicetnunc/gh-pages/html-preview/sw.js'): The user denied permission to use Service Worker.
````

Trying to mint then led to the transaction being backtracked.

**Comment by pichiste** [comment]
> The preview is still not working for me at all. Blank screen with this error:
> 
> 
> 
> ````
> 
> Uncaught (in promise) DOMException: Failed to register a ServiceWorker for scope ('https://hicetnunc2000.github.io/hicetnunc/gh-pages/html-preview/') with script ('https://hicetnunc2000.github.io/hicetnunc/gh-pages/html-preview/sw.js'): The user denied permission to use Service Worker.
> 
> ````
> 
> 
> 
> Trying to mint then led to the transaction being backtracked.

Which browser? Can you try hard refresh?

**Comment by pichiste** [comment]
@codewithfeeling The only place I can reproduce this is on Brave, is that what you're using? If so, disabling the shield for the site fixes the problem:

![Screen Shot 2021-03-31 at 7 14 42 AM](https://user-images.githubusercontent.com/111979/113094126-7326ca80-91f1-11eb-8d86-edae79315e95.png)

![Screen Shot 2021-03-31 at 7 15 02 AM](https://user-images.githubusercontent.com/111979/113094132-75892480-91f1-11eb-90e9-a66511469c92.png)

**Comment by codewithfeeling** [comment]
I am indeed using Brave. Thanks for the info re shields - will do that and confirm.

**Comment by codewithfeeling** [comment]
Turning Shields off fixes the issue on Brave - thanks.

**Comment by pichiste** [comment]
@crzypatchwork you can close this issue, thanks
